extensions/slack-messages/src/slackClient.ts

```diff
@@ -0,0 +1,283 @@
+
+    if (!conversationsJSON) {
+      return undefined;
+    }
+    const conversations = (JSON.parse(conversationsJSON) as ConversationInfo[]).filter((c) => c.hasUnread);
+    conversations.forEach((c) => {
+      c.lastMessageDate = new Date(c.lastMessageDate as unknown as string);
+      c.messages?.forEach((m) => {
+        m.date = new Date(m.date as unknown as string);
+      });
+    });
+    localConversations = conversations;
+    if (conversations) {
+      const viewModel = this.viewModelForConversations(conversations);
+      viewModel.isLoading = true;
+      return viewModel;
+    }
+    return undefined;
+  }
+
+  async useListViewModel(): Promise<ListViewModel> {
+    const conversations = await this.conversations();
+    if (conversations) {
+      return this.viewModelForConversations(conversations.filter((c) => c.hasUnread));
+    }
+
+    return {
+      isLoading: false,
+    };
+  }
+  async conversations(): Promise<ConversationInfo[] | undefined> {
+    const conversations = await this.fetchConversations();
+    if (conversations) {
+      this.cacheConversations(conversations);
+    }
+    localConversations = conversations;
+    return conversations;
+  }
+
+  /**
+   * # Fuck Slack's API.
+   * Don't make calls to this method very often. When needing to mark a read position, a client should set a timer before making the call.
+   * In this way, any further updates needed during the timeout will not generate extra calls (just one per channel). This is useful for when reading scroll-back history, or following a busy live channel.
+   * A timeout of 5 seconds is a good starting point.
+   */
+  async markAsRead(conversation: ConversationInfo): Promise<Error | undefined> {
+    if (conversation.id && conversation.lastMessageTs && conversation.lastMessageDate) {
+      if (conversation.hasUnread) {
+        try {
+          const read = await web.conversations.mark({ channel: conversation.id, ts: conversation.lastMessageTs });
+          return new Error(read.error);
+        } catch (error) {
+          return error as Error;
+        }
+      }
+    }
+    return undefined;
+  }
+
+  async markAllAsRead(): Promise<void> {
+    localConversations
+      ?.filter((c) => c.hasUnread)
+      ?.forEach(async (c) => {
+        try {
+          //TODO: switch type of convers and mark im.mark, conversations.mark, groups.mark
+          const read = await web.conversations.mark({ channel: c.id, ts: c.lastMessageTs });
+          return new Error(read.error);
+        } catch (error) {
+          return error as Error;
+        }
+      });
+  }
+
+  private async fetchConversations(): Promise<ConversationInfo[] | undefined> {
+    try {
+      const conversations = await web.conversations.list({
+        exclude_archived: true,
+        types: "public_channel, private_channel, mpim, im",
+      });
+      const fuckingChannels = conversations.channels?.filter((channel) => channel.is_member);
```

Comment:
Language 😅

Reply:
Ahaha, oh yeah, I remember how I couldn't get the channels that I really needed and annoyed by the Slack's API 😁 
Thanks for the pointing out!